src/ArcherContainer.js

```diff
@@ -154,21 +154,22 @@ export class ArcherContainer extends React.Component<Props, State> {
   };
 
   registerTransition = (fromElement: string, relation: RelationType): void => {
-    // TODO prevent duplicate registering
     // TODO improve the state merge... (should be shorter)
-    const fromTo = [...this.state.fromTo];
-    const newFromTo = {
-      ...relation,
-      from: { ...relation.from, id: fromElement },
-    };
-    fromTo.push(newFromTo);
-
-    this.setState((currentState: State) => ({
+    const { fromTo } = this.state;
+    const newFromTo = [
+      ...(!fromTo.find(sd => sd.from.id === fromElement) ? [{
```

Comment:
Does it really work when the element you are registering is the source of multiple arrows?

In that case `fromTo.find(sd => sd.from.id === fromElement)` would return something (because of the first arrow), so we wouldn't register the new relation.

Does it make sense?

(btw, I would've added an additional variable like `const existingRelationWithSameSource = fromTo.find(sd => sd.from.id === fromElement)` so that the big array right below becomes more understandable 😉 )

Reply:
Ah yes, I didn't think about multiple arrows with the same source.
So I guess it should check if `relationExists` with the same `from` and `to`. But, what if an anchor changes? I suppose there shouldn't be multiple relations with same `from` and `to` but different anchors? Or maybe that also has to be supported for some rare use cases.

And yes I did think about using an extra variable, especially to add an extra error message in case the relation already exists, but then also wanted to make it shorter since the comment at the top states "state merge should be shorter". But that might be referring to `this.setState()` with the callback function.